### PR TITLE
UNO-203 new event to handle ng delivery id. Fixed test to phpunit8

### PR DESCRIPTION
**Pull request by Yury-Razhkou**
related to: UNO-203
Added new event to handle NG data in webhooks.
Fixed triggering DeliveryCreatedEvent.
Fixed\updated unit tests to phpunit8.

### CODE REVIEW

File: model/events/DeliveryPublicationCreatedEvent.php [hunk 1]
```diff
@@ -29,18 +29,22 @@ class DeliveryPublicationCreatedEvent implements Event
...
      * @param string $deliveryClassUri
+     * @param string $testUri
      */
-    public function __construct($deliverTenantId, $publicationId, $deliveryClassUri)
+    public function __construct($deliverTenantId, $publicationId, $deliveryClassUri, $testUri)
     {
         $this->deliverTenantId = $deliverTenantId;
         $this->publicationId = $publicationId;
         $this->deliveryClassUri = $deliveryClassUri;
+        $this->testUri = $testUri;
```
Comment (ktsirkou) [inline model/events/DeliveryPublicationCreatedEvent.php:47]: Why is this added here?

Comment (Yury-Razhkou) [reply model/events/DeliveryPublicationCreatedEvent.php:47]: In CG delivery created without testUri. This field needed in several places (ex. DeliveryCreatedEvent in webhook functionality) and breaks code.

File: model/events/NGDeliveryCreatedEvent.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+namespace oat\taoDeliverConnect\model\events;
+
+use core_kernel_classes_Resource;
+use oat\tao\model\webhooks\WebhookSerializableEventInterface;
+use \core_kernel_persistence_Exception;
+use oat\taoDeliverConnect\model\delivery\factory\RemoteDeliveryFactory;
+use oat\taoDeliveryRdf\model\DeliveryAssemblyService;
+use oat\taoDeliveryRdf\model\event\AbstractDeliveryEvent;
+
+class NGDeliveryCreatedEvent extends AbstractDeliveryEvent implements WebhookSerializableEventInterface
```
Comment (wazelin) [inline model/events/NGDeliveryCreatedEvent.php:31]: I thought we're just gonna extend the existing webhook with a new property.
Has it been decided to go with a new event?

File: test/unit/model/PublicationServiceTest.php [hunk 3]
```diff
@@ -63,7 +63,7 @@ public function testPublish()
         /** @var RestApiClient|MockObject $apiClient */
         $apiClient = $this->getMockBuilder(RestApiClient::class)
```
Comment (wazelin) [inline test/unit/model/PublicationServiceTest.php:64]: TBF the whole thing seems to be replaceable with `$this->createMock()`.

File: model/events/NGDeliveryCreatedEvent.php [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            'delivery' => $this->delivery->getUri(),
+        ];
+    }
+
+    /**
+     * @return string
+     */
+    public function getWebhookEventName()
+    {
+        return 'NGDeliveryCreatedEvent';
```
Comment (Alroniks) [inline model/events/NGDeliveryCreatedEvent.php:75]:
```suggestion
        return __CLASS__;
```

File: model/events/NGDeliveryCreatedEvent.php (hunk 2 above)
Comment (Alroniks) [inline model/events/NGDeliveryCreatedEvent.php:31]: Better to avoid using the narrowly known abbreviations for open source code. One solution is `NextGenDeliveryCreatedEvent`, but I would try to find a more reliable name.

File: model/events/NGDeliveryCreatedEvent.php (hunk 4 above)
Comment (Alroniks) [inline model/events/NGDeliveryCreatedEvent.php:75]: Or use 'getName' method if it should just get event name.

File: model/events/NGDeliveryCreatedEvent.php (hunk 2 above)
Comment (Yury-Razhkou) [reply model/events/NGDeliveryCreatedEvent.php:31]: DeliveryCreatedEvent is used in different environments and we can't add field that exists only on NG

File: model/tasks/LookupForDeliveryTask.php [hunk 5]
```diff
@@ -126,17 +135,17 @@ public function __invoke($params = [])
      * @throws GuzzleException
      * @throws InconsistentDataException
      */
-    private function handleSuccess(LookupSnapshot $lookupSnapshot, DeliverTenant $deliverTenant, $deliveryClassUri)
+    private function handleSuccess(LookupSnapshot $lookupSnapshot, DeliverTenant $deliverTenant, $deliveryClassUri, $testUri)
     {
         $deliveryClass = $this->getClass($deliveryClassUri);
-
+        $testResource = $this->getResource($testUri);
```
Comment (Alroniks) [inline model/tasks/LookupForDeliveryTask.php:141]: The test can be removed while publishing is happening on the nextgen env, and it will raise a fatal error. Needed to handle such situations.

File: model/events/NGDeliveryCreatedEvent.php (hunk 2 above)
Comment (wazelin) [reply model/events/NGDeliveryCreatedEvent.php:31]: Oh yeah, that's true.
Thanks!

File: test/unit/model/PublicationServiceTest.php (hunk 3 above)
Comment (Yury-Razhkou) [reply test/unit/model/PublicationServiceTest.php]: Thanks. Replaced where it was possible.

File: model/tasks/LookupForDeliveryTask.php (hunk 5 above)
Comment (Yury-Razhkou) [reply model/tasks/LookupForDeliveryTask.php]: Refactored to use a string variable.

File: model/events/NGDeliveryCreatedEvent.php (hunk 4 above)
Comment (Yury-Razhkou) [reply model/events/NGDeliveryCreatedEvent.php]: Thanks. Applied this variant

File: model/events/NGDeliveryCreatedEvent.php (hunk 2 above)
Comment (Yury-Razhkou) [reply model/events/NGDeliveryCreatedEvent.php:31]: I don't like current naming too. But `NextGen..` soon will be not `NextGen..` and this will also confusing. Something like `RemoteDeliveryCreated...` is also confusing because we have separate functionality for remote publishing.

**Comment by wazelin** [comment]
Do we have a follow-up task in order to configure the event listeners?

**Comment by Yury-Razhkou** [comment]
> Do we have a follow-up task in order to configure the event listeners?

Webhook configuration required. But we do not have endpoints.

File: model/events/NGDeliveryCreatedEvent.php [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+     * @throws core_kernel_persistence_Exception
+     */
+    public function serializeForWebhook()
+    {
+        $testProperty = new \core_kernel_classes_Property(DeliveryAssemblyService::PROPERTY_ORIGIN);
+        $ngDeliveryProperty = new \core_kernel_classes_Property(RemoteDeliveryFactory::PROPERTY_PUBLISHED_DELIVERY_ID);
+        return [
+            'deliveryId' => $this->deliveryUri,
+            'testId' => $this->delivery->getOnePropertyValue($testProperty)->getUri(),
+            'ngDeliveryId' => $this->delivery->getOnePropertyValue($ngDeliveryProperty),
```
Comment (wazelin) [inline model/events/NGDeliveryCreatedEvent.php:89]:
```suggestion
            'ngDeliveryId' => (string)$this->delivery->getOnePropertyValue($ngDeliveryProperty),
```
Otherwise it'll return an object, and the actual value will be in its `literal` property.

File: model/tasks/LookupForDeliveryTask.php [hunk 7]
```diff
@@ -177,12 +193,14 @@ private function createDeliveryResource($label, LookupSnapshot $lookupSnapshot,
...
             RemoteDeliveryFactory::PROPERTY_TENANT_ID => $deliverTenant->getId(),
             RemoteDeliveryFactory::PROPERTY_PUBLISHED_DELIVERY_ID => $lookupSnapshot->getPublishedDeliveryId(),
+            DeliveryAssemblyService::PROPERTY_ORIGIN => $testUri
         ];
 
         $deliveryResource = $deliveryClass->createInstanceWithProperties($properties);
 
         $eventManager = $this->getServiceLocator()->get(EventManager::SERVICE_ID);
         $eventManager->trigger(new DeliveryCreatedEvent($deliveryResource));
+        $eventManager->trigger(new NGDeliveryCreatedEvent($deliveryResource));
```
Comment (SergiiTao) [inline model/tasks/LookupForDeliveryTask.php:203]: whet is the difference between these two events?

Comment (SergiiTao) [reply model/tasks/LookupForDeliveryTask.php:203]: I see. serializeForWebhook slightly different. :)

Comment (Yury-Razhkou) [reply model/tasks/LookupForDeliveryTask.php:203]: The main reason to add new event was an additional field that not exists in common tao event. And we can't add it there because it exists only in udir.

File: model/events/NGDeliveryCreatedEvent.php [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        return $this->getName();
+    }
+
+    /**
+     * @return array
+     * @throws core_kernel_persistence_Exception
+     */
+    public function serializeForWebhook()
+    {
+        $testProperty = new \core_kernel_classes_Property(DeliveryAssemblyService::PROPERTY_ORIGIN);
```
Comment (SergiiTao) [inline model/events/NGDeliveryCreatedEvent.php:84]: isn't it better to ingest already prepared values to avoid calling DB without necessity?

File: model/events/NGDeliveryCreatedEvent.php (hunk 2 above)
Comment (Alroniks) [reply model/events/NGDeliveryCreatedEvent.php:28]: I suggest in this case `ExternalDeliveryCreatedEvent`. It is truth and looks better than NG.

File: model/events/DeliveryPublicationLookupAttemptEvent.php [hunk 9]
```diff
@@ -35,20 +35,26 @@ class DeliveryPublicationLookupAttemptEvent implements Event
...
     /**
      * DeliveryPublicationCreatedEvent constructor.
      * @param string $deliverTenantId
      * @param string $publicationId
      * @param string $deliveryClassUri
      * @param integer $currentAttempt
+     * @param string $testUri
      */
-    public function __construct($deliverTenantId, $publicationId, $deliveryClassUri, $currentAttempt)
+    public function __construct($deliverTenantId, $publicationId, $deliveryClassUri, $currentAttempt,$testUri)
```
Comment (Alroniks) [inline model/events/DeliveryPublicationLookupAttemptEvent.php:51]: Seems formatting lost a bit. Space needed after the comma.

File: model/events/NGDeliveryCreatedEvent.php [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+class NGDeliveryCreatedEvent extends AbstractDeliveryEvent implements WebhookSerializableEventInterface
+{
+    /**
+     * @var string
+     */
+    private $testUri;
+    /**
+     * @var string
+     */
+    private $ngDeliveryUri;
```
Comment (Alroniks) [inline model/events/NGDeliveryCreatedEvent.php:37]:
```suggestion
    private $externalDeliveryUri;
```

File: model/events/NGDeliveryCreatedEvent.php [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+
+    /**
+     * @return array
+     */
+    public function serializeForWebhook()
+    {
+        return [
+            'deliveryId' => $this->deliveryUri,
+            'testId' => $this->testUri,
+            'ngDeliveryId' => $this->ngDeliveryUri,
```
Comment (Alroniks) [inline model/events/NGDeliveryCreatedEvent.php:88]: Is it should be with ng prefix? Could it be changed to external too?

File: model/tasks/LookupForDeliveryTask.php [hunk 12]
```diff
@@ -194,12 +214,12 @@ private function createDeliveryResource($label, LookupSnapshot $lookupSnapshot,
      * @param integer $attempt
      * @return Report
      */
-    private function handleUnready($deliverTenantId, $publicationId, $deliveryClassUri, $attempt)
+    private function handleUnready($deliverTenantId, $publicationId, $deliveryClassUri, $attempt,$testUri)
```
Comment (Alroniks) [inline model/tasks/LookupForDeliveryTask.php:217]: Seems formatting lost a bit. Space needed after the comma.

File: model/tasks/LookupForDeliveryTask.php [hunk 13]
```diff
@@ -194,12 +214,12 @@ private function createDeliveryResource($label, LookupSnapshot $lookupSnapshot,
...
     {
         sleep($this->getPublicationService()->getLookupInterval());
         /** @var EventManager $em */
         $em = $this->getServiceLocator()->get(EventManager::class);
-        $em->trigger(new DeliveryPublicationLookupAttemptEvent($deliverTenantId, $publicationId, $deliveryClassUri, $attempt));
+        $em->trigger(new DeliveryPublicationLookupAttemptEvent($deliverTenantId, $publicationId, $deliveryClassUri, $attempt,$testUri));
```
Comment (Alroniks) [inline model/tasks/LookupForDeliveryTask.php:222]: Seems formatting lost a bit. Space needed after the comma.

File: model/events/NGDeliveryCreatedEvent.php (hunk 11 above)
Comment (ktsirkou) [reply model/events/NGDeliveryCreatedEvent.php:88]: or `remoteDeliveryId`

Comment (Yury-Razhkou) [reply model/events/NGDeliveryCreatedEvent.php:88]: For me `remoteDeliveryId` associated with other functionality - remote publishing. It is also remote publishing but the same naming will be confusing

File: model/events/NGDeliveryCreatedEvent.php (hunk 8 above)
Comment (Yury-Razhkou) [reply model/events/NGDeliveryCreatedEvent.php]: Done

File: model/events/DeliveryPublicationLookupAttemptEvent.php (hunk 9 above)
Comment (Yury-Razhkou) [reply model/events/DeliveryPublicationLookupAttemptEvent.php]: Fixed

File: model/events/NGDeliveryCreatedEvent.php (hunk 10 above)
Comment (Yury-Razhkou) [reply model/events/NGDeliveryCreatedEvent.php]: renamed al `ng...` to `external...`

File: model/tasks/LookupForDeliveryTask.php (hunk 12 above)
Comment (Yury-Razhkou) [reply model/tasks/LookupForDeliveryTask.php]: Fixed

File: model/tasks/LookupForDeliveryTask.php (hunk 13 above)
Comment (Yury-Razhkou) [reply model/tasks/LookupForDeliveryTask.php]: Fixed
